plugins/MumblePluginProcess.hpp

```diff
@@ -0,0 +1,313 @@
+// Copyright 2020 The Mumble Developers. All rights reserved.
+// Use of this source code is governed by a BSD-style license
+// that can be found in the LICENSE file at the root of the
+// Mumble source tree or at <https://www.mumble.info/LICENSE>.
+
+#ifndef MUMBLEPLUGINPROCESS_HPP_
+#define MUMBLEPLUGINPROCESS_HPP_
+
+#include "mumble_plugin_utils.h"
+
+#include <chrono>
+#include <cstdint>
+#include <cstring>
+#include <map>
+#include <vector>
+
+#ifdef OS_LINUX
+#	include <sstream>
+
+#	include <sys/uio.h>
+#endif
+
+typedef uint32_t procid_t;
+typedef uint64_t procptr_t;
+
+/// Abstract class.
+/// Only defines stuff that can be used with both Linux and Windows processes.
+class MumblePluginProcess {
+protected:
+	bool m_ok;
+	procid_t id;
+	std::string name;
+	uint8_t pointerSize;
+
+public:
+	inline bool isOk() { return m_ok; }
+#ifdef OS_WINDOWS
+	virtual bool peek(const procptr_t address, void *dst, const size_t size) = 0;
+#else
+	bool peek(const procptr_t address, void *dst, const size_t size) {
+		iovec in;
+		in.iov_base = reinterpret_cast< void * >(address);
+		in.iov_len  = size;
+
+		iovec out;
+		out.iov_base = dst;
+		out.iov_len  = size;
+
+		const auto ret = process_vm_readv(id, &out, 1, &in, 1, 0);
+
+		return (ret != -1 && static_cast< size_t >(ret) == in.iov_len);
+	}
+#endif
```

Comment:
Just make `peek` a pure virtual function and put the platform-specific code in the respective sub-class. That way it'll be easier to find the implementation for a given function (and tidies up the code a bit).

And besides: Implementations shouldn't be in header files (for non-templates)

Reply:
The reason why I didn't do it is due to `peek()`'s implementation being the same for both native and Wine processes.

Reply:
"native" being "nativ on Linux"? Then I don't see a problem with moving the implementation to the linux class :thinking:

Reply:
Yes, native Linux processes. The method is the same for processes that are handled by Wine.

I would have to copy the exact same implementation to the Windows class if I moved it to the Linux one.

Reply:
Does wine use a windows instance?

Reply:
From a memory viewpoint yes, you see the process exactly as it is on Windows.

When the game is running through Wine `MumblePluginProcessWindows` is used.

Reply:
> When the game is running through Wine MumblePluginProcessWindows is used.

Yeah that was what I wanted to know :+1:

Okay in that case, leave the implementation in the base-class (though as I said elsewhere I would move the actual implementation into a separate cpp file) but make it virtual. And the overwrite it in the Windows class. That way you still don't need the preprocessor stuff. That only makes the code harder to parse :)

Reply:
Any ideas for a name for the file? I was thinking about `ProcessLinuxCommon.cpp` or `ProcessCommonLinux.cpp`.

Since it doesn't define a new class, also `Process_linux.cpp` or `Process_Linux.cpp` could do.

Reply:
As far as I understood we'd end up with 3 files & classes: `Process` (base-class), `WindowsProcess` and `LinuxProcess` :thinking:

Reply:
`Process` (`Process.cpp/.h`), `ProcessWindows` (`ProcessWindows.cpp/.h`) and `ProcessLinux` (`ProcessLinux.cpp/.h`).

You suggested to remove the preprocessor stuff and move the implementation into a separate `.cpp` file, how should that file be called?

Reply:
That should be `Process.cpp` ("separate" in this case means "a different file than the header file")

Reply:
Ah, I see.

Reply:
Unfortunately this cannot work, as you can see from the CI failure.

A preprocessor statement is still required in order to mark the virtual functions as pure for Windows.

Reply:
I ended up moving the WinAPI implementation to `Process`, which I think is the best solution.

Now `ProcessWindows` only refers to the process itself, rather than the OS.

Reply:
I have to say that I am not a big fan of using preprocessor directives everywhere to divide the implementation into chunks. That is just terrible when it comes to readability of the code.
I much prefer using inheritance to solve this issue. If you don't want to have virtual functions in the resulting code, you can apply a small preprocessor trick for choosing the correct type when actually using these classes, but for the implementation this should be separated by having virtual functions which are overriden as needed

Reply:
https://travis-ci.com/github/mumble-voip/mumble/builds/191018159

https://github.com/mumble-voip/mumble/compare/9df6e9d1571844b0e567e416ff699f8b94d54880..1057ca833c39324e4f7339aa8832f32d481c52b1

Reply:
The error your encountered should be easily fixable: https://stackoverflow.com/questions/3065154/undefined-reference-to-vtable

Just define all functions which are not pure virtual and the error should go away :thinking: 

> Right now the only idea I have for removing the preprocessor directives in a relatively clean way is to move the OS-specific functions from Process.cpp to two separate files (one for each OS). Then we can include them at the bottom of Process.cpp.

Well that'd at least be better than having the preprocessor directives all over the place.

> I thought you had the same idea in #4536 (comment).

No. I was really talking about using virtual functions and inheritance. No preprocessor involved (except for finally choosing which one to instantiate).
Basically what you started to implement in the linked diff. That was exactly what I was talking about.

Reply:
Could you write an example, please?

Reply:
You said that the `peekProc` was the same for Windows and Wine processes, right? In that case you'd have:

Process.h:
```
class Process {
public:
    virtual bool peek(const procptr_t address, void *dst, const size_t size);
};
```
Process.cpp:
```
bool Process::peek(const procptr_t address, void *dst, const size_t size) {
    // implementation windows/wine
}
```

ProcessLinux.h
```
class ProcessLinux {
public:
    virtual bool peek(const procptr_t address, void *dst, const size_t size) override;
};
```

ProcessLinux.cpp
```
bool Process::peek(const procptr_t address, void *dst, const size_t size) {
    // implementation Linux
}
```

And then when it actually comes to instantiating one or the other, we could do something like this:
```
#ifdef OS_WINDOWS
    using process_t = ProcessWIndows;
#else 
    using process_t = ProcessLinux
#endif 

    process_t proc(/* whatever arguments it needs */);
```

Or given that even on Linux we might have to use `ProcessWindows` we have to actually use OO polymorphism:
```
    std::unique_ptr<Process> proc;
    if (windows || wine) {
        proc = std::make_unique<ProcessWindows>(/* whatever arguments it needs */);
    } else {
        proc = std::make_unique<ProcessLinux>(/* whatever arguments it needs */);
    };
```

Reply:
`peek()`'s implementation is completely different for Linux (Wine included) and Windows.

Reply:
Well in that case the function's definition would be pure virtual in `Process` and would then be overwritten by both `ProcessWindows` and `ProcessLinux`

Reply:
It can't be done, as I said the two classes are chosen depending on the process rather than the OS.

Reply:
???
Why exactly can't it be done? It doesn't matter how you choose the classes. You can always encapsulate different implementations by overwriting existing functionality.

Please provide an actual description of what the problems exactly are (an example with one function might be enough) and how you think to solve it

Reply:
![Process](https://user-images.githubusercontent.com/5897523/97053326-055fde00-1583-11eb-9c3f-6df7e9d2bda3.png)

On Linux `ProcessLinux` is used when the target process is native. If the process runs through Wine, `ProcessWindows` is used.

On Windows only `ProcessWindows` is used.

Reply:
So the issue ultimately comes down to this:
The `peek` function uses a different implementation on Linux that it does on Windows, regardless of which process is being inspected (wine vs native)?
But on the other hand we do have the `exportedSymbol` function that depends on whether an actual linux process is being inspected vs a windows process (potentially via wine), regardless of the actual operating system it is running on?

Reply:
Exactly!

Reply:
Okay I see.

In that case we simply need another layer of indirection:
We start from our current baseline `Process` class that declares thee `peek` and `exportedSymbol` functions as pure virtual.

Then we have 2 classes that could be called `LinuxHost` and `WindowsHost` that'll implement the `peek` function as required for the respective OS.

And finally we have our well known `ProcessLinux` and `ProcessWIndows` classes that implement the `exportedSymbol` function.

The `*Host` classes inherit from `Process` while the inheritance of the `Process*` classes will depend on the OS that is being compiled on. For that we could do something like this:
```
#ifdef OS_WINDOWS
using host_class = WindowsHost;
#else 
using host_class = LinuxHost;
#endif 

class ProcessLinux : public host_class {
};
```

With this solution we still have some preprocessor statements in our code but in this way I think the code remains much more readable than by doing it in the current way :thinking: 

----

Btw: What about MacOS? Do we even support the overlay on MacOS?

Reply:
We support the overlay on macOS: https://github.com/mumble-voip/mumble/blob/master/overlay_gl/init_mac.c

Regarding positional audio plugins, only Link is currently supported: https://github.com/mumble-voip/mumble/blob/master/plugins/link/link-posix.cpp

In theory we can read the memory through `vm_read()` and get the address of an exported symbol by parsing the Mach-O header: https://www.mikeash.com/pyblog/friday-qa-2012-11-30-lets-build-a-mach-o-executable.html

About the classes: your idea is excellent!